semimpl/do_test.go

```diff
@@ -0,0 +1,177 @@
+package semimpl_test
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/benchttp/runner/semimpl"
+)
+
+func TestDo(t *testing.T) {
+	t.Run("stop until maxIter is reached", func(t *testing.T) {
+		var (
+			numWorkers = 1
+			gotIter    = 0
+			maxIter    = 10
+			expIter    = 10
+		)
+
+		semimpl.Do(context.Background(), numWorkers, maxIter, func() {
+			gotIter++
+		})
+
+		if gotIter != expIter {
+			t.Errorf("iterations: exp %d, got %d", expIter, gotIter)
+		}
+	})
+
+	t.Run("stop on context timeout", func(t *testing.T) {
+		var (
+			timeout    = 50 * time.Millisecond
+			interval   = timeout / 5
+			numWorkers = 1
+			maxIter    = 0 // infinite iterations
+
+			margin      = 15 * time.Millisecond
```

Comment:
[question]
How did you decide how much should the margin be?
This fails on my computer, I need at least to replace this by 20 for this test to pass.

Reply:
100% empirically, I'll update it.